src/shared/remoted_op.c

```diff
@@ -188,45 +143,108 @@ int parse_agent_update_msg (char *msg,
+                            char **config_sum,
+                            char **merged_sum,
+                            char **agent_ip,
+                            char **labels)
+{
+    char *msg_tmp = NULL;
+    char *str_tmp = NULL;
+    char *line = NULL;
+    char sdelim[] = { '\n', '\0' };
+    const char * AGENT_IP_LABEL = "#\"_agent_ip\":";
 
-    os_strdup(msg_tmp, *uname);
+    // Temporary coping the msg string
+    os_strdup(msg, msg_tmp);
 
-    // Get merged.mg sum
-    str_tmp = end_line + 1;
-    if (*str_tmp != '\"' && *str_tmp != '!' && (end_line = strchr(str_tmp, ' '), end_line)) {
-        *end_line = '\0';
-        end_line++;
+    for (line = strtok(msg_tmp, sdelim); line; line = strtok(NULL, sdelim)) {
+        switch (*line) {
+        case '#':  // System label
+        case '!':  // Hidden label
+        case '\"': // Regular label
+            // The _agent_ip will not be appended to the labels string.
+            // Instead it will be returned in the agent_ip parameter.
+            if(!strncmp(line, AGENT_IP_LABEL, strlen(AGENT_IP_LABEL))) {
+                os_strdup(line + strlen(AGENT_IP_LABEL), *agent_ip);
+            }
+            else {
+                wm_strcat(labels, line, '\n');
+            }
+            break;
+        default:
+            // uname - wazuh version / config sum
+            if (str_tmp = strstr(line, " / "), str_tmp)
+            {
+                *str_tmp = '\0';
+                str_tmp += 3;
+                os_strdup(str_tmp, *config_sum);
 
-        if (strncmp(end_line, SHAREDCFG_FILENAME "\n", sizeof(SHAREDCFG_FILENAME "\n")-1) != 0) {
-            *merged_sum = NULL;
-        }
-        else {
-            os_strdup(str_tmp, *merged_sum);
-        }
-    }
-    else { // If we didn't find merged.mg we should keep the end line
-        end_line = str_tmp;
-    }
+                if (str_tmp = strstr(line, " - "), str_tmp){
+                    *str_tmp = '\0';
+                    str_tmp += 3;
+                    os_strdup(str_tmp, *version);
+                }
 
-    // Get the agent ip
-    const char * AGENT_IP_TAG = "#\"_agent_ip\":";
-    str_tmp = end_line;
+                parse_uname_string(line, os_name, os_major, os_minor, os_build,
+                                   os_version, os_codename, os_platform, os_arch);
 
-    end_line = strchr(str_tmp, '\n') + 1;
-    if (end_line && !strncmp(end_line, AGENT_IP_TAG, strlen(AGENT_IP_TAG))) {
-        os_strdup(end_line + strlen(AGENT_IP_TAG), *agent_ip);
+                os_strdup(line, *uname);
+            }
+            // merged sum
+            else if (str_tmp = strchr(line, ' '), str_tmp)
+            {
+                *str_tmp = '\0';
+                str_tmp++;
 
-        if (end_line = strchr(*agent_ip, '\n'), end_line){
-            *end_line = '\0';
+                if (strncmp(str_tmp, SHAREDCFG_FILENAME, strlen(SHAREDCFG_FILENAME)-1) != 0) {
+                    *merged_sum = NULL;
```

Comment:
Shouldn't we expand this logic of guaranteeing NULL if no value was found to be parsed? Maybe we can set everything to NULL at the begining, previous to parsing.

Reply:
DONE